### ISSUE TITLE
"Virtual functions are pure" build error with RLock (HXCPP)

### ISSUE THREAD

**Issue by EliteMasterEric** [vegardit/haxe-concurrent#21]
The following error occurs during compilation of any project which includes `hx.concurrent.lock.RLock`, even a minimal one.,

```
Error: ./src/hx/concurrent/lock/RLock.cpp: In static member function ‘static Dynamic hx::concurrent::lock::RLock_obj::__CreateEmpty()’:
./src/hx/concurrent/lock/RLock.cpp:47:49: error: invalid new-expression of abstract class type ‘hx::concurrent::lock::RLock_obj’
   47 | Dynamic RLock_obj::__CreateEmpty() { return new RLock_obj; }
      |                                                 ^~~~~~~~~
In file included from ./src/hx/concurrent/lock/RLock.cpp:13:
include/hx/concurrent/lock/RLock.h:21:30: note:   because the following virtual functions are pure within ‘hx::concurrent::lock::RLock_obj’:
   21 | class HXCPP_CLASS_ATTRIBUTES RLock_obj : public  ::hx::concurrent::lock::AbstractAcquirable_obj
      |                              ^~~~~~~~~
In file included from ./src/hx/concurrent/lock/RLock.cpp:7:
include/hx/concurrent/lock/AbstractAcquirable.h:48:30: note:     ‘virtual bool hx::concurrent::lock::AbstractAcquirable_obj::tryAcquire(int)’
   48 |                 virtual bool tryAcquire(int timeoutMS) = 0;
      |                              ^~~~~~~~~~
./src/hx/concurrent/lock/RLock.cpp: In static member function ‘static Dynamic hx::concurrent::lock::RLock_obj::__Create(hx::DynamicArray)’:
./src/hx/concurrent/lock/RLock.cpp:53:65: error: invalid new-expression of abstract class type ‘hx::concurrent::lock::RLock_obj’
   53 |         ::hx::ObjectPtr< RLock_obj > _hx_result = new RLock_obj();
      |                                                                 ^
./src/hx/concurrent/lock/RLock.cpp: In static member function ‘static hx::ObjectPtr<hx::concurrent::lock::RLock_obj> hx::concurrent::lock::RLock_obj::__new()’:
./src/hx/concurrent/lock/RLock.cpp:211:61: error: invalid new-expression of abstract class type ‘hx::concurrent::lock::RLock_obj’
  211 |         ::hx::ObjectPtr< RLock_obj > __this = new RLock_obj();
      |                                                             ^
./src/hx/concurrent/lock/RLock.cpp: In static member function ‘static void hx::concurrent::lock::RLock_obj::__register()’:
./src/hx/concurrent/lock/RLock.cpp:356:19: error: cannot declare variable ‘_hx_dummy’ to be of abstract type ‘hx::concurrent::lock::RLock_obj’
  356 |         RLock_obj _hx_dummy;
      |                   ^~~~~~~~~
```

Building with HXCPP on the Linux platform.

**Comment by sebthom** [comment]
You can check the tests and the github action which successful builds on Linux targeting HXCPP

**Comment by sebthom** [comment]
Sorry didn't plan to close the issue. Fat fingers

**Comment by EliteMasterEric** [comment]
Apologies for the delayed response. After running the test provided by the Github Action (and finding it successfully builds and runs), I did some more investigation. I found that the issue occurs when building the project using the Lime framework's build tool, and was able to recreate the problem in a minimal sample project.

You can find my full writeup at #22.

**Comment by joshtynjala** [comment]
I just discovered that I also needed to add `-dce full` to compile haxe-concurrent with hxcpp on Linux. I should note that I am not using haxe-concurrent in a Lime project. It's just a regular Haxe project with a normal .hxml file.

Luckily, I can use `-dce full` in this project without any issues. However, not everyone can be so strict. That makes me wonder, even if there's an underlying hxcpp bug, what can haxe-concurrent do to work around this issue and support projects without `-dce full`?
